src/mixins/java/org/spongepowered/common/mixin/core/entity/player/ServerPlayerEntityMixin.java

```diff
@@ -468,26 +495,144 @@ public void teleportTo(net.minecraft.world.server.ServerWorld world, double x, d
+                    });
+                    // Sponge End
+                    // This is done here (it's not in Forge) to ensure that it's done. This is to ensure
+                    // advancements are triggered.
+                    this.shadow$triggerDimensionChangeTriggers(serverworld);
+                    // Sponge Start: make sure the right object was returned
+                    if (e != (Object) this) {
+                        throw new IllegalArgumentException(String.format("Teleporter %s "
+                                + "did not return the expected player entity: got %s, expected PlayerEntity %s", spongeTeleporter, e, this));
+                    }
+
+                    // If we need to reposition: well... reposition.
+                    // Downside: portals won't come with us, but with how it's implemented in Forge,
+                    // not sure how we'd do this
+                    final Vector3d destination = VecHelper.toVector3d(this.shadow$position());
+                    final ChangeEntityWorldEvent.Reposition reposition = SpongeEventFactory.createChangeEntityWorldEventReposition(
+                            PhaseTracker.getCauseStackManager().getCurrentCause(),
+                            (org.spongepowered.api.entity.Entity) this,
+                            (org.spongepowered.api.world.server.ServerWorld) serverworld,
+                            currentPosition,
+                            destination,
+                            (org.spongepowered.api.world.server.ServerWorld) originalDestinationWorld,
+                            originalDestination,
+                            (org.spongepowered.api.world.server.ServerWorld) targetWorld
+                    );
+                    if (reposition.isCancelled()) {
+                        // send them back to the original destination
+                        this.shadow$setPos(originalDestination.getX(), originalDestination.getY(), originalDestination.getZ());
+                    } else if (reposition.getDestinationPosition() != destination) {
+                        final Vector3d finalPosition = reposition.getDestinationPosition();
+                        this.shadow$setPos(finalPosition.getX(), finalPosition.getY(), finalPosition.getZ());
+                    }
+                    // Sponge End
+                    this.gameMode.setLevel(targetWorld);
+                    this.connection.send(new SPlayerAbilitiesPacket(this.abilities));
+                    playerlist.sendLevelInfo((ServerPlayerEntity) (Object) this, targetWorld);
+                    playerlist.sendAllPlayerInfo((ServerPlayerEntity) (Object) this);
+
+                    for (EffectInstance effectinstance : this.shadow$getActiveEffects()) {
+                        this.connection.send(new SPlayEntityEffectPacket(this.shadow$getId(), effectinstance));
+                    }
+
+                    this.connection.send(new SPlaySoundEventPacket(1032, BlockPos.ZERO, 0, false));
```

Comment:
This should only be called if it is actually a portal else API usages of NetherPortalType will have the swoosh sound which would be odd imo.

Reply:
Assuming this code is copied to a util method for handling post teleport logic.